godot-core/src/engine/script_instance.rs

```diff
@@ -318,32 +254,99 @@ pub fn create_script_instance<T: ScriptInstance>(rs_instance: T) -> *mut c_void
     }
 }
 
+/// Ref-guard for a mutable reference to T where T implements `ScriptInstance`.
+pub struct SiMut<'a, T: ScriptInstance> {
+    mut_ref: &'a mut T,
+    cell: Pin<&'a GdCell<T>>,
+    base_ref: &'a Base<T::Base>,
+}
+
+impl<'a, T: ScriptInstance> SiMut<'a, T> {
+    fn new(
+        cell: Pin<&'a GdCell<T>>,
+        cell_guard: &'a mut MutGuard<T>,
+        base_ref: &'a Base<T::Base>,
+    ) -> Self {
+        let mut_ref = cell_guard.deref_mut();
+
+        Self {
+            mut_ref,
+            cell,
+            base_ref,
+        }
+    }
+}
+
+impl<'a, T: ScriptInstance> Deref for SiMut<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.mut_ref
+    }
+}
+
+impl<'a, T: ScriptInstance> DerefMut for SiMut<'a, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.mut_ref
+    }
+}
+
+impl<'a, T: ScriptInstance> SiMut<'a, T> {
+    pub fn base(&self) -> BaseRef<T> {
+        BaseRef::new(self.base_ref.to_gd(), self.mut_ref)
+    }
+
+    pub fn base_mut(&mut self) -> BaseMut<T> {
+        let guard = self.cell.make_inaccessible(self.mut_ref).unwrap();
+
+        BaseMut::new(self.base_ref.to_gd(), guard)
+    }
+}
+
 mod script_instance_info {
     use std::any::type_name;
-    use std::cell::{BorrowError, Ref, RefMut};
     use std::ffi::c_void;
     use std::mem::ManuallyDrop;
+    use std::ops::Deref;
+    use std::pin::Pin;
+
+    use godot_cell::{GdCell, RefGuard};
 
     use crate::builtin::{GString, StringName, Variant};
     use crate::engine::ScriptLanguage;
     use crate::obj::Gd;
     use crate::private::handle_panic;
     use crate::sys;
 
-    use super::{ScriptInstance, ScriptInstanceData};
-
-    fn borrow_instance_mut<T: ScriptInstance>(instance: &ScriptInstanceData<T>) -> RefMut<'_, T> {
-        instance.inner.borrow_mut()
+    use super::{ScriptInstance, ScriptInstanceData, SiMut};
+
+    macro_rules! borrow_panic {
+        ($type:ty) => {
+            |err| {
+                panic!(
+                    "\
+                    ScriptInstance borrow failed, already bound; T = {}.\n  \
+                    Make sure to use `SiMut::base_mut()` when possible.\n  \
+                    Details: {err}.\
+                ",
+                    type_name::<$type>(),
+                )
+            }
+        };
```

Comment:
This could be a generic function instead of a macro.
Also no need for the closure.